policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+
+    def init_from_json(self, json):
+        self.kind = KIND_ENDPOINTS
+        self.name = json["metadata"]["name"]
+        self.namespace = json["metadata"]["namespace"]
+        self.subsets = json["subsets"]
+
+    @property
+    def key(self):
+        return "%s/%s" % (self.namespace, self.name)
+
+    def generate_svc_profiles_pods(self):
+        """
+        Endpoints objects contain a list of subsets
+        containing pod and policy info
+        subsets: [
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+        ]
+        :return: A generated dict of profile-pod associations
+        :rtype: a dict of profiles mapping to a list of assoc pods
+        """
+        def verify_args(ports_obj, try_arg):
+            try:
+                arg = ports_obj[try_arg]
+            except KeyError:
+                arg = None
+            return arg
+
+        self.association_list = {}
+        for subset in self.subsets:
+            # Base profile_name string
+            profile_name = "%s_svc_%s" % (self.namespace, self.name)
+
+            # Generate a list of new svc profiles per port spec
+            port_rules = []
+            for port_spec in subset["ports"]:
+                dst_port = verify_args(port_spec, "port")
+                protocol = verify_args(port_spec, "protocol").lower()
+                if protocol and dst_port:
+                    port_rules.append(Rule(action="allow",
+                                           dst_ports=[dst_port],
+                                           protocol=protocol))
+                    profile_name = "%s_%s%s" % (
+                        profile_name, protocol, dst_port)
+
+            # Create Profile for the subset
+            if not _datastore_client.profile_exists(profile_name):
+                _log.info("Creating Profile %s" % profile_name)
+                profile_path = PROFILE_PATH % {"profile_id": profile_name}
+                _datastore_client.etcd_client.write(
+                    profile_path + "tags", '["%s"]' % profile_name)
+
+                # Determine rule set
+                default_allow = Rule(action="allow")
+                rules = Rules(id=profile_name,
+                              inbound_rules=port_rules,
+                              outbound_rules=[default_allow])
+
+                # Write rules to profile
+                _datastore_client.etcd_client.write(
+                    profile_path + "rules", rules.to_json())
+            else:
+                _log.warning("Profile %s already exists" % profile_name)
+
+            # Generate list of pod names
+            pods = []
```

Comment:
We need to be resilient to the targetRef not existing - this won't exist for all services.

If it doesn't exist its fine - we just don't apply policy on those endpoints - but we definitely can't crash due to a `KeyError`